### PR TITLE
Introduce @flatten annotation

### LINKED ISSUE #623: Provide Rust serde-like flatten behavior for flattening/sharing/capturing fields (500USD Bounty)
---

**From the maintainer Li Haoyi**: I'm putting a 500USD bounty on this issue, payable by bank transfer on a merged PR implementing this.

---

Sometimes when processing JSON you want to use a shared structure across various models. The serde crate in Rust wanted to provide support for this feature and provided it through the [flatten](https://noritada.github.io/attr-flatten.html) attribute. 
It would be very useful if upickle also supported this.

Lets say we have an API which  supports pagination as a general concept and both our **users** and **products** endpoints support pagination.

## Usecase: Sharing via Flattening 

```json
{
  "limit": 100,
  "offset": 200,
  "total": 1053,
  "users": [
    {"id": "49824073-979f-4814-be10-5ea416ee1c2f", "username": "john_doe"},
    ...
  ]
}
```

```json
{
  "limit": 50,
  "offset": 100,
  "total": 2103,
  "users": [
    {"id": "abf24073-979f-4814-be10-5ea416ee1c2f", "shortname": "blender", "description":"It slices it dices."},
    ...
  ]
}
```

It would be useful if we could use the following models to capture this.

```scala
case class Pagination(limit:Int, offset:Int, total:Int)
case class User(id:String, username:String)
case class Users(users:List[User], @upickle.implicits.flatten pagination:Pagination)

case class ProductInfo(id:String, shortname:String, description:String)
case class Products(products:List[Product], @upickle.implicits.flatten pagination:Pagination)
```

## Usecase: Extra Fields

This can also be used as a mechanism for handling extra fields like the Rust crate also does.

```json
{
   "name":"myLibrary",
   "devDependencies": {
        "vite": "^5.4.2",
        "typescript":"^5.0.0"   
    },
   "nonStandardField1":"blah",
   "customDependencies": {
       "@myEnterprise/enterpriseLib": "1.0.0"
   }
}
```

```scala
case class PackageManifest(
    name:String, 
    devDependencies:Map[String,String], 
    @upickle.implicits.flatten otherStuff:Map[String, ujson.Value]
}
```

### PR DESCRIPTION
**Pull request by nox213**
closes #623 

This PR introduces the @flatten annotation.

The following limitations apply:
1.	Flattening more than two Maps is not supported. (I believe supporting multiple Maps when deriving a Reader offers little benefit.)
2.	Type parameters do not seem to be properly resolved in the following scenario:
```
 case class Param[T](@flatten t: T)
object Param {
   implicit def rw[T]: RW[Param[T]] = upickle.default.macroRW // compile error
  implicit def rw[SomeClass]: RW[Param[SomeClass]] = upickle.default.macroRW // works
}
3. When using the @flatten annotation on a Map, the key type must be String. (Allowing user-defined case classes as keys introduces a high risk of key conflicts, so I’ve restricted it to String.)


Question About Test Cases
The following test case that I added does not pass on Scala 2.12.18:
```
    test("flattenWithMap") {
      import Flatten._
      val value = HasMap(Map("key1" -> "value1", "key2" -> "value2"), 10)
      rw(value, """{"key1":"value1","key2":"value2","i":10}""")
    }
```
The reason is as follows:
```
 val a = scala.collection.mutable.Map[String, String]
a += "key1" -> "value1"
a += "key2" -> "value2"
```

When using a mutable Map, the output varies by Scala version:
• Scala 2.12.18:  `Map("key2" -> "value2", "key1" -> "value1")`
• Scala 2.13.11:   `Map("key1" -> "value1", "key2" -> "value2")`
• Scala 3.3.4:      `Map("key1" -> "value1", "key2" -> "value2")`

Only Scala 2.12.18 produces the reversed order. Since a mutable Map is used internally in the Reader to support the @flatten annotation, serializing, deserializing, and then serializing again results in a binary format mismatch.

Do you have any suggestions for a good solution? Sorting the Map during read and write operations for @flatten was considered, but it seems too costly in terms of performance, so it wasn’t implemented.

### CODE REVIEW

File: upickle/implicits/src-2/upickle/implicits/internal/Macros.scala [hunk 1]
```diff
@@ -268,109 +300,150 @@ object Macros {
...
         .annotations
-        .find(_.tpe == typeOf[upickle.implicits.allowUnknownKeys])
-        .flatMap(_.scalaArgs.headOption)
+        .find(_.tree.tpe == typeOf[upickle.implicits.allowUnknownKeys])
+        .flatMap(_.tree.children.tail.headOption)
         .map { case Literal(Constant(b)) => b.asInstanceOf[Boolean] }
-        .headOption
 
-      val defaults = deriveDefaults(companion, hasDefaults)
+      val allFields = fields.flatMap(_.allFields).filter(_.flatten != Flatten.Map)
```
Comment (nox213) [inline upickle/implicits/src-2/upickle/implicits/internal/Macros.scala:320]: When flattening a Map, I handled this by not assigning it an index (using -1) and treating it as an exception. And, it is not included in the field count. This approach ensures that when key-value pairs for the flattened Map are not found in the JSON, it is considered valid and does not trigger a check in checkErrorMissingKeys. (ie, it's a just empty Map)

Another approach I considered was assigning an index to the flattened Map field and using storeValueIfNotFound to create an empty Map in such cases. Which approach do you think is better?

**Comment by lihaoyi** [comment]
@nox213 this looks great. Could you update the PR description with a summary of the implemented semantics, with examples, as well as a summary of the implementation strategy? That would make the PR much easier to review and merge

**Comment by nox213** [comment]
Sure, I’ll tag you when I’m ready.

File: upickle/implicits/src-2/upickle/implicits/internal/Macros.scala [hunk 2]
```diff
@@ -33,43 +33,13 @@ object Macros {
           s"[[${clsSymbol.name}]]. This may be due to a bug in scalac (SI-7567) " +
           "that arises when a case class within a function is upickle. As a " +
           "workaround, move the declaration to the module-level."
-        Left(msg)
-      }else{
+        fail(tpe, msg)
```
Comment (nox213) [inline upickle/implicits/src-2/upickle/implicits/internal/Macros.scala:36]: Looking at its usage, errors are not being accumulated, and there didn’t seem to be a reason to use Either, so I modified it to throw an error immediately.

**Comment by nox213** [comment]
@lihaoyi I polished the code a little and updated the PR.

**Comment by lihaoyi** [comment]
@nox213 at I high level, would it be possible to implement this without the `case class Field` refactor? If we could preserve the existing method signatures mostly-in-place, we could add forwarders and maintain binary compatibility which would allow us to release this without a breaking major version

**Comment by lihaoyi** [comment]
I guess more generally I'd like to see whether we can land this without breaking binary compatibility

**Comment by lihaoyi** [comment]
I think this could work:

* `currentKey`/`storeToMap` variables can be made binary compatible by moving them from the upstream trait into the various downstream impls

* The macro changes could be made binary compatible by leaving the existing macros in place but deprecated, and just making a `Macros2` class containing all the new implementations

**Comment by lihaoyi** [comment]
For the Scala 2 vs Scala 3 approach, could we use the Scala 3 approach in Scala 2 as well? It's best to keep the two implementations mostly consistent, so future changes can be done more easily across both versions

**Comment by lihaoyi** [comment]
> ```scala 
> case class Param[T](@flatten t: T)
> object Param {
>    implicit def rw[T]: RW[Param[T]] = upickle.default.macroRW // compile error when this function is called to derive instance
>   implicit val rw[SomeClass]: RW[Param[SomeClass]] = upickle.default.macroRW // works
> }
> ```

Can this be fixed by making it `implicit def rw[T: RW]`?

**Comment by lihaoyi** [comment]
> When using the @flatten annotation on a Map, the type of key must be String. (Allowing user-defined case classes as keys introduces a high risk of key conflicts, so I’ve restricted it to String.) Also, only immutable.Map is supported. This makes derivation of Reader simpler.
>
> @flatten is only allowed to immutable.Map. This makes derivation of Reader simpler.

How hard would it be to generalize this to other collection types and key types? 

* One use case for other collection types would be to support `Seq[(String, T)]` or `LinkedHashMap[String, T]` for cases where people want to preserve ordering
* uPickle already supports non-String map keys via https://com-lihaoyi.github.io/upickle/#JSONDictionaryFormats, so it would be nice if we could be consistent here as well when flattening out maps

**Comment by lihaoyi** [comment]
Left some high level comments. It's a pretty invasive change so I expect it will take some time to discuss the changes, first at a high-level and then down to the lines of code being changed, just to set expectations.

**Comment by nox213** [comment]
@lihaoyi  Thanks for your review.

> Can this be fixed by making it implicit def rw[T: RW]?
I'm not sure whether it's possible. I will investigate more on it.

> How hard would it be to generalize this to other collection types and key types?
I believe this is doable. For other collection types, I could store the type of collection that is flatten and then use that type to construct it from Reader. For keys, I didn't do it because I thought that's not the use cases. I will try this. 

> For the Scala 2 vs Scala 3 approach, could we use the Scala 3 approach in Scala 2 as well? It's best to keep the two implementations mostly consistent, so future changes can be done more easily across both versions
Sure, I will use the Scala 3 approach in Scala 2 (no `case class Field`).

**Comment by nox213** [comment]
The challenging part seems to be bincompat. Would the best approach be to implement @flatten without changing the existing function signatures then? And if changing the function signatures is unavoidable, the next best option would be to keep the existing functions (marking them as deprecated) and implement separate macros (e.g., a new Macro class in Scala 2). In this case, would it make sense to introduce a conditional branch in specific functions to decide whether to call the deprecated macro or the new one? For example, determining this based on the presence of the @flatten annotation. Did I understand this correctly?

One thing I’m curious about: as I understand it, macro code is only called internally, so I assumed it would be safe in terms of bincompat. However, can even functions that are not called externally impact bincompat?

**Comment by lihaoyi** [comment]
For bincompat, we could leave the existing `Macros` file completely untouched and just make a new copy `Macros2` that we use going forward. No need to conditionally branch on the presence of `@flatten`, the new code works great. We just need to leave the old code in place for bincompat in case anyone downstream is using it (probably nobody, but good to be careful in preserving bincompat anyway)

**Comment by lihaoyi** [comment]
> I believe this is doable. For other collection types, I could store the type of collection that is flatten and then use that type to construct it from Reader. For keys, I didn't do it because I thought that's not the use cases. I will try this.

I think what you would need to do is take the type of the collection, resolve the `Reader` for it, and use that `Reader` instance at runtime in order to construct the collection from the individual key-value pairs

For keys, I think you should be able to do the same thing: take the type, resolve the `Reader`, and use the `Reader` to construct the key objects from the JSON keys and then pass those to the collection `Reader`

**Comment by nox213** [comment]
Thanks for your explanation and tips.

To summarize the direction I need to take for the changes:
1. For Scala 2, keep the existing Macros.scala unchanged and create a new Macros2.scala to apply the modifications (for bincompat), and the new code will use Macros2.scala.
2. Move the definitions of `currentKey` and `storeToMap` from the upstream trait to the downstream implementation (for bincompat).
3. Update the implementation to support flattening for collection types other than immutable.Map.
4. Allow Map to have key types other than String.
5. For the Scala 2 macro implementation, aim to follow a design as similar as possible to Scala 3, without using case class Field.
 - In the current Scala 2 implementation, all the information necessary for derivation (e.g., symbols, types, mappedName, etc.) is generated upfront and passed to wrapCaseN for construction. If I were to adopt an approach similar to Scala 3, wrapCaseN would instead start with just the type being derived and then call a function to fetch the field information for that type.

Is this the right direction to proceed?

Lastly, for Macros2.scala, would it be better to apply private or [package] private as much as possible if those restrictions can be applied?

**Comment by lihaoyi** [comment]
@nox213 yes that's right. Making the new macros package private would indeed help, though some of the supporting runtime APIs may need to remain public so we won't be able to privatize everything

**Comment by nox213** [comment]
Okay, I will work on this.

**Comment by nox213** [comment]
@lihaoyi
I have some questions regarding bincompat of Scala3.
1. I modified types of some fields(`construct`, `visitor0`) in `CaseClassReader3`. Do you have some ideas to keep bincompat? Is it okay to make a copy of `CaseClassReader3` `like CaseClassReader3V2` and use it instead?
2. In macros, is it okay to maintain bincompat by just keeping old methods and making them deprecated?

**Comment by lihaoyi** [comment]
> @lihaoyi I have some questions regarding bincompat of Scala3.
> 
> 1. I modified types of some fields(`construct`, `visitor0`) in `CaseClassReader3`. Do you have some ideas to keep bincompat? Is it okay to make a copy of `CaseClassReader3` like `CaseClassReader3V2` and use it instead?

Making a copy should work

> 2. In macros, is it okay to maintain bincompat by just keeping old methods and making them deprecated?

Yes leaving the old methods deprecated should work as well

**Comment by nox213** [comment]
@lihaoyi 
[77a882e](https://github.com/com-lihaoyi/upickle/pull/642/commits/77a882e751bb24ce9abc65ecba19cbf327880e26)
This is the change that move `storeToMap` and `currentKey` to downstream trait. Since `storeToMap` should be used in `visitValue`, I had to overwrite this function in downstream trait. What do you think of making a copy of `ObjectContext.scala` instead so that we don't need to define `visitValue` separately for scala 2 & 3.

**Comment by lihaoyi** [comment]
@nox213 I don't quite understand what you are asking

**Comment by nox213** [comment]
If I define `storeToMap` not inside of  `BaseCaseObjectContext` but inside of the trait that is generated by macro codes, I have to override `visitValue` as well.

**Comment by nox213** [comment]
Okay, I will keep it as it is

**Comment by nox213** [comment]
@lihaoyi 
I made Scala2 macros works with Iterable[(Key, Value)] (removing restrictions to only work with Map and String keys). But I ran into one problem when I worked on Scala 3: the stricter quote system does not allow me to use the same approach with Scala 2.

First, for deriving a Writer of collection with non-string key, I did like below for Scala 2.
```scala
val writeKey = if (keyType =:= typeOf[String]) q"key" else q"upickle.default.write(key)(implicitly[${c.prefix}.Writer[$keyType]])"
                q"""
                  $select.foreach { case (key, value) =>
                  this.writeSnippetMappedName[R, $valueType](
                    ctx,
                    $writeKey,
                    implicitly[${c.prefix}.Writer[$valueType]],
                    value
                    )
                  }
                """
```
Basically, I summoned Writer[Key] and called `upickle.default.write` with that Writer to convert key value to string.
But this doesn't work for Scala 3 since quote of Scala 3 do not allow me to generate such a code. Scala 3 requires that all elements referenced in the quote are valid and resolvable in the scope of the quote itself if I understood correctly. Do you have some ideas that convert key to string without calling `write`?

```scala
'{
                ${select.asExprOf[Iterable[(keyTpe, valueTpe)]]}.foreach { (k, v) =>
                  ${self}.writeSnippetMappedName[R, valueTpe](
                    ${ctx},
                    upickle.default.write(key)(summonInline[keyWriterTpe]), // this doesn't compile
                    summonInline[writerTpe],
                    v,
                  )
                }
              }
```

**Comment by lihaoyi** [comment]
@nox213 I'm not actually sure on the Scala 3 side. @jodersky do you have any ideas how this can be done with Scala 3 metaprogramming?

**Comment by nox213** [comment]
What I tried was that
```scala
val writeRef = Ref(Symbol.requiredMethod("upickle.default.write"))
          (keyTpe0.asType, valueTpe0.asType, keyWriterTpe0.asType, valueWriterTpe0.asType) match {
            case ('[keyTpe], '[valueTpe], '[keyWriterTpe], '[valueWriterTpe]) =>
              val typeAppliedWrite = TypeApply(writeRef, List(TypeTree.of[keyTpe]))
              val snippet =
                if (keyTpe0 =:= TypeRepr.of[String])
                 ... // be ommitted
                else
                  '{
                      ${select.asExprOf[Iterable[(keyTpe, valueTpe)]]}.foreach { (k, v) =>
                        ${self}.writeSnippetMappedName[R, valueTpe](
                          ${ctx},
                          ${Apply(
                              Apply(
                                typeAppliedWrite,
                                List('{k}.asTerm, '{-1}.asTerm, '{false}.asTerm, '{false}.asTerm)
                              ),
                              List(
                                '{summonInline[keyWriterTpe]}.asTerm
                              )
                            )
                            .asExprOf[String]},
                          summonInline[valueWriterTpe],
                          v,
                        )
                      }
                    }
```
I used `requiredMethod` to access the method symbol out of the scope. But the error was that 
```
[error]     |               Found:    (upickle.Flatten.KeyClass.rw :
[error]     |                 upickle.default.ReadWriter[upickle.Flatten.KeyClass])
[error]     |               Required: Api.this.Writer[upickle.Flatten.KeyClass]
``` 
Those two types are treated as different type because of their path is different. I got stuck at this point.

**Comment by nox213** [comment]
@lihaoyi 

> > ```scala
> > case class Param[T](@flatten t: T)
> > object Param {
> >    implicit def rw[T]: RW[Param[T]] = upickle.default.macroRW // compile error when this function is called to derive instance
> >   implicit val rw[SomeClass]: RW[Param[SomeClass]] = upickle.default.macroRW // works
> > }
> > ```
> 
> Can this be fixed by making it `implicit def rw[T: RW]`?

I misunderstood your comment. I tested with your fix but it still doesn't work.
```scala
  object FlattenTestWithType {
    implicit def rw[T: RW]: RW[FlattenTestWithType[T]] = upickle.default.macroRW // compile error
  }
```

I’ve pushed the changes to support collections other than Map (iterable flattening is now possible) so far. [a07839d](https://github.com/com-lihaoyi/upickle/pull/642/commits/a07839dc9b4c0c7a9ebffac2f32861d55e4751f8)

**Comment by jodersky** [comment]
I just saw your ping now. Let me know if there's still something I can help with

**Comment by nox213** [comment]
@jodersky Thanks, could you check theses [comment](https://github.com/com-lihaoyi/upickle/pull/642#issuecomment-2541505944), [comment](https://github.com/com-lihaoyi/upickle/pull/642#issuecomment-2545686020)?
Basically, I tried to make flatten annotation works for collections that has non-string key. To support it, I need to generate code that calls functions defined in `upickle.default` which lie outside the scope from `macros.scala`.

**Comment by lihaoyi** [comment]
Sure! Would be happy to merge it first 

Give me some time to properly review it. Still probably need @jodersky 's help to look through some of the scala 3 macro stuff, and I'll give it a pass as well

**Comment by lihaoyi** [comment]
I read through and think it looks reasonable.

One request: can we have some tests for the edge cases and failure modes of `@flatten`? Some I can think of:

* What if the flattened field itself contains a field which collides with the enclosing case class?
* Can we flatten two fields in a single case class list?
* If we flatten two fields in a case class, what if they have fields which collide?
* What if we `@flatten` a field whose type type that cannot be flattened?
* What if we `@flatten` a collection of invalid type (e.g. doesn't contain a string key?)

I don't really have any opinion on the exact semantics in these situations, but regardless of what we choose we should have unit tests asserting their success or their failure and error message

**Comment by nox213** [comment]
> What if the flattened field itself contains a field which collides with the enclosing case class? 
If we flatten two fields in a case class, what if they have fields which collide?

It fails with following errors. Before process the flatten annotation, macro checks whether there are duplicated `mappendName`.
```
[error] /Users/devsisters/repos/upickle/upickle/test/src/upickle/MacroTests.scala:225:48: There are multiple fields with the same key.
[error] Following keys are duplicated: x.
[error]     implicit val rw: RW[Bar] = upickle.default.macroRW
[error]                                                ^
[error] one error found
```

> Can we flatten two fields in a single case class list?

Yes, this is possible, but my implementation only supports flattening multiple case classes and a single collection at a time.

> What if we @flatten a field whose type that cannot be flattened?

```
[error] /Users/devsisters/repos/upickle/upickle/test/src/upickle/MacroTests.scala:221:48: Invalid type for flattening: Int.
[error]     implicit val rw: RW[Foo] = upickle.default.macroRW
[error]                                                ^
[error] one error found
```

> What if we @flatten a collection of invalid type?

This also fails with an same error above.

I added a few more tests to cover the parts that were not addressed by the tests I had previously created ([2eb1ab8](https://github.com/com-lihaoyi/upickle/pull/642/commits/2eb1ab8c0e22cbbb90e02533ee4d3883605c6016)).

**Comment by nox213** [comment]
@lihaoyi While adding some tests, I thought of a case where a collection could lead to runtime collisions. For example:
```scala
case class Foo(i: Int, @flatten m: Map[String, String])
```
If m = Map("i" -> "value"), there would be a key collision, but this can only be detected at runtime. Referring to the behavior of the Rust library linked in the issue, here’s how it handles such cases:
When writing, it generates JSON like this:
```json
{"i": 1, "i": "value"}
```
And when reading it back, it fails with an error like:
```
Error("duplicate field `i`", line: 1, column: 16)
```
On the other hand, if the map field appears first, like this:
```scala
case class Foo(@flatten m: Map[String, String], i: Int)
```
The generated JSON would look like this:
```json
{"i": "value", "i": 1}
```
In this case, when trying to read it back, it would result in a runtime error while attempting to convert "value" to an integer.

In upickle, it behaves almost the same except for the first case.
When it reads,
```
{"i": 1, "i": "value"}
```
my implementation produces `Foo(1, Map.empty)`.
However, making it raise an error when there are duplicate keys at runtime seems doable.

- Do you think it would be better to raise an error in such cases rather than silently succeeding?
- A runtime key collision can cause an error when reading a value and attempting a type conversion. Would it be acceptable to keep this behavior? I currently can’t think of a better way to handle it.

**Comment by lihaoyi** [comment]
I think the current behavior for the case you mentioned is reasonable, but would it be possible to make it raise an error on write? That would probably be more intuitive to users than generating JSON a dictionary with duplicate keys

**Comment by nox213** [comment]
I made it raise an error on write an added an failure test case as well. ([e4071f6](https://github.com/com-lihaoyi/upickle/pull/642/commits/e4071f6b9a1967d39dababacf417480cce05b6f5))
```
X upickle.MacroTests.runtimeCollision 26ms 
  java.lang.Exception: Key collision detected for the following keys: i, s
    upickle.Flatten$Foo$$anon$67.write0(MacroTests.scala:221)
    upickle.Flatten$Foo$$anon$67.write0(MacroTests.scala:221)
    upickle.core.Types$Writer.write(Types.scala:143)
```

**Comment by samikrc** [comment]
I have limited experience in Scala 3 macros, so probably not a great reviewer of this issue, however posing some general questions:
- There seem to be a few classnames like this: `CaseClassReader3V2` etc? Would it be better to change the code for the existing class itself? The git system maintains our versions, so we do not need to introduce V2 etc?
- The new method `CaseClassReader3` starts with a `@deprecated` tag. Is that expected?

**Comment by nox213** [comment]
> There seem to be a few classnames like this: CaseClassReader3V2 etc? Would it be better to change the code for the existing class itself? The git system maintains our versions, so we do not need to introduce V2 etc?

It's a temporary name that will be changed to `CaseClassReader3`  later.

> The new method CaseClassReader3 starts with a @deprecated tag. Is that expected?

It's a class that existed before which is replaced by `CaseClassReader3V2 `

File: upickle/implicits/src-3/upickle/implicits/macros.scala [hunk 3]
```diff
@@ -221,59 +374,92 @@ def tagKeyImpl[T](using Quotes, Type[T])(thisOuter: Expr[upickle.core.Types with
...
+            typeSymbol.typeRef.dealias.asType match {
+              case '[t] =>
+                val newTpe = TypeRepr.of[t]
+                val (term, nextOffset) = newTpe match {
+                  case t: AppliedType => apply(newTpe, t.args, i)
+                  case t: TypeRef => apply(newTpe, List.empty, i)
+                  case t: TermRef => (Ref(t.classSymbol.get.companionModule), i)
+                }
+                (term :: terms, nextOffset)
+              case _ =>
```
Comment (jodersky) [inline upickle/implicits/src-3/upickle/implicits/macros.scala:423]: do you need this catch-all case? Wouldn't ``[t]` already match everything?

**Review by jodersky — COMMENTED** [review]
The Scala 3 macro code looks reasonable. I also appreciate that you cleaned up a bunch of stuff and restricted visibilities of internal methods! I do have some comments on them however, mostly minor issues and questions to help my own understanding.

I also second Haoyi, in that I can't guarantee that there aren't any subtle bugs, but we can use the test suite for that. Speaking of, do we have some kind of performance tests we can use to weed out regressions? From what I see in the code, most flatten-specific macro parts are behind a flag (if the annotation is present), so I don't *think* there will be any regressions, but it's always good to test this stuff.

---

Regarding supporting of non-string keys: I think it's probably best to tackle that in a separate PR later on too; this one is already quite large.

To your [specific question](https://github.com/com-lihaoyi/upickle/pull/642#issuecomment-2541505944), @nox213

```scala
'{
                ${select.asExprOf[Iterable[(keyTpe, valueTpe)]]}.foreach { (k, v) =>
                  ${self}.writeSnippetMappedName[R, valueTpe](
                    ${ctx},
                    upickle.default.write(key)(summonInline[keyWriterTpe]), // this doesn't compile
                    summonInline[writerTpe],
                    v,
                  )
                }
              }
```

one thing you could try is take the outer "API trait" as a parameter (similar to what is done in `writeSnippets(..., inline thisOuter: upickle.core.Types with upickle.implicits.MacrosCommon...)`), and use some typecasting if necessary.

Maybe something like

```scala
def foo(api: Expr[upickle.core.Types with upickle.implicits.MacrosCommon]) =
  ${outer}.write(key.asInstanceOf[Any])(summonInline[keyWriterTpe])
```

File: upickle/implicits/src-3/upickle/implicits/macros.scala [hunk 4]
```diff
@@ -114,83 +196,154 @@ def keyToIndexImpl[T](x: Expr[String])(using Quotes, Type[T]): Expr[Int] = {
...
+            case '[t] => Expr(allFields[t].map(_._2).toSet)
+          }
+          val writerTpe0 = TypeRepr.of[W].appliedTo(valueTpe0)
+          (valueTpe0.asType, writerTpe0.asType) match {
+            case ('[valueTpe], '[writerTpe])=>
+              val snippet = '{
+                val collisions = ${select.asExprOf[Iterable[(String, valueTpe)]]}.map(_._1).toSet.intersect(${allKeysExpr})
+                if (collisions.nonEmpty) {
+                  throw new Exception("Key collision detected for the following keys: " + collisions.mkString(", "))
+                }
```
[the remark this answers is not in the document]
Comment (nox213) [reply upickle/implicits/src-3/upickle/implicits/macros.scala:285]: `select` refers to a collection, and we check whether its keys of elements conflict with existing keys. It seems we cannot perform this check at compile time, since we don’t know what the collection contains until runtime.

File: upickle/implicits/src-3/upickle/implicits/macros.scala [hunk 5]
```diff
@@ -221,59 +374,92 @@ def tagKeyImpl[T](using Quotes, Type[T])(thisOuter: Expr[upickle.core.Types with
...
+                val (term, nextOffset) = newTpe match {
+                  case t: AppliedType => apply(newTpe, t.args, i)
+                  case t: TypeRef => apply(newTpe, List.empty, i)
+                  case t: TermRef => (Ref(t.classSymbol.get.companionModule), i)
+                }
+                (term :: terms, nextOffset)
+              case _ =>
+                report.errorAndAbort(s"Unsupported type $typeSymbol for flattening")
             }
+          } else report.errorAndAbort(s"${typeSymbol} is not a case class or a immutable.Map")
```
[the remark this answers is not in the document]
Comment (nox213) [reply upickle/implicits/src-3/upickle/implicits/macros.scala]: [64a5048](https://github.com/com-lihaoyi/upickle/pull/642/commits/64a504803151219fae8b54972ffb89987fb80f28)

File: upickle/implicits/src-3/upickle/implicits/macros.scala [hunk 6]
```diff
@@ -389,3 +576,96 @@ def defineEnumVisitorsImpl[T0, T <: Tuple](prefix: Expr[Any], macroX: String)(us
...
+  import quotes.reflect._
+  tpe <:< TypeRepr.of[Iterable[(String, _)]]
+}
+
+private def isCaseClass(using Quotes)(typeSymbol: quotes.reflect.Symbol): Boolean = {
+  import quotes.reflect._
+  typeSymbol.isClassDef && typeSymbol.flags.is(Flags.Case)
+}
+
+private def extractKeyValueTypes(using Quotes)(tpe: quotes.reflect.TypeRepr): (quotes.reflect.TypeRepr, quotes.reflect.TypeRepr, quotes.reflect.TypeRepr) =
```
[the remark this answers is not in the document]
Comment (nox213) [reply upickle/implicits/src-3/upickle/implicits/macros.scala]: [fc02f23](https://github.com/com-lihaoyi/upickle/pull/642/commits/fc02f238ac4210858a3db2bb35c780bd442dc3bb)
Yes, you're right. My intention was that extract K, V from collections like
```scala
Map[K, V]
Iterable[(K, V)]
```

File: upickle/implicits/src-3/upickle/implicits/macros.scala [hunk 7]
```diff
@@ -60,28 +62,110 @@ def extractIgnoreUnknownKeysImpl[T](using Quotes, Type[T]): Expr[List[Boolean]]
...
+            fieldLabelsImpl0[t]
+              .flatMap { case (rawLabel, label) =>
+                val newDefaults = getDefaultParamsImpl0[t]
+                val newClassTypeRepr = TypeRepr.of[T]
+                loop(rawLabel, label, newClassTypeRepr, newDefaults)
+              }
+          case _ =>
+            report.errorAndAbort(s"Unsupported type $typeSymbol for flattening")
+        }
+      } else report.errorAndAbort(s"${typeSymbol} is not a case class or a immutable.Map")
```
[the remark this answers is not in the document]
Comment (nox213) [reply upickle/implicits/src-3/upickle/implicits/macros.scala]: [379aab3](https://github.com/com-lihaoyi/upickle/pull/642/commits/379aab37746dad5782f54411bdccd96296687f9e)

File: upickle/implicits/src-3/upickle/implicits/macros.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (nox213) [reply upickle/implicits/src-3/upickle/implicits/macros.scala]: [966022d](https://github.com/com-lihaoyi/upickle/pull/642/commits/966022dd648b246dab61eceb23d4656938a315f6)

**Comment by nox213** [comment]
@jodersky Thanks for the reviews! I addressed your reviews.

> one thing you could try is take the outer "API trait" as a parameter (similar to what is done in writeSnippets(..., inline thisOuter: upickle.core.Types with upickle.implicits.MacrosCommon...)), and use some typecasting if necessary.

This can only be done when there are no cyclic dependencies, right? If I understand correctly, the project that defines the ‘API trait’ depends on upickle.implicits.

**Comment by lihaoyi** [comment]
Thanks @nox213! I have merged the PR and tagged version 4.1.0 which should go out in the next hour or two

Please send me an email with your international bank transfer details so I can send you the bounty payment

**Comment by nox213** [comment]
Thank you! I've sent the required information via email. I hope no major bugs arise.
